### PR TITLE
Faster ClassNameHelper

### PR DESCRIPTION
**Pull request by staabm**
70% faster edge case in `vendor/bin/phpunit tests/PHPStan/Analyser/AnalyserIntegrationTest.php --filter testBug10979` - which is currently the slowest `AnalyserIntegrationTest` in the repo.

Before this PR
```
➜  phpstan-src git:(regr) ✗ php vendor/bin/phpunit tests/PHPStan/Analyser/AnalyserIntegrationTest.php --filter testBug10979 
PHPUnit 9.6.23 by Sebastian Bergmann and contributors.

Random Seed:   1747292998
Warning:       No code coverage driver available

.                                                                   1 / 1 (100%)

Time: 00:03.944, Memory: 74.00 MB

OK (1 test, 1 assertion)
➜  phpstan-src git:(regr) ✗ php vendor/bin/phpunit tests/PHPStan/Analyser/AnalyserIntegrationTest.php --filter testBug10979
```

After this PR
```
➜  phpstan-src git:(regr) ✗ php vendor/bin/phpunit tests/PHPStan/Analyser/AnalyserIntegrationTest.php --filter testBug10979
PHPUnit 9.6.23 by Sebastian Bergmann and contributors.

Random Seed:   1747293054
Warning:       No code coverage driver available

.                                                                   1 / 1 (100%)

Time: 00:01.452, Memory: 74.00 MB

OK (1 test, 1 assertion)
```

---

Profile before the PR

grafik

triggered on [`IntersectionType->getFiniteTypes()`](https://github.com/phpstan/phpstan-src/blob/2fb3dbe482362e96ba2f9e23768d627a80064ae8/src/Type/IntersectionType.php#L1210-L1227)
```
➜  phpstan-src git:(fast) ✗ php -v
PHP 8.3.21 (cli) (built: May  6 2025 13:58:10) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.3.21, Copyright (c) Zend Technologies
```

### CODE REVIEW

File: src/Reflection/ClassNameHelper.php [hunk 1]
```diff
@@ -3,15 +3,23 @@
...
+	/** @var array<string, bool> */
+	private static array $checked = [];
+
 	public static function isValidClassName(string $name): bool
 	{
-		// from https://stackoverflow.com/questions/3195614/validate-class-method-names-with-regex#comment104531582_12011255
-		return Strings::match(ltrim($name, '\\'), '/^[a-zA-Z_\x80-\xff][a-zA-Z0-9_\x80-\xff]*(\\\\[a-zA-Z_\x80-\xff][a-zA-Z0-9_\x80-\xff]*)*$/') !== null;
+		if (!array_key_exists($name, self::$checked)) {
+			// from https://stackoverflow.com/questions/3195614/validate-class-method-names-with-regex#comment104531582_12011255
+			self::$checked[$name] = Strings::match(ltrim($name, '\\'), '/^[a-zA-Z_\x80-\xff][a-zA-Z0-9_\x80-\xff]*(\\\\[a-zA-Z_\x80-\xff][a-zA-Z0-9_\x80-\xff]*)*$/') !== null;
```
Comment (mvorisek) [inline src/Reflection/ClassNameHelper.php:19]: How much faster https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/v3.75.0/src/DocBlock/TypeExpression.php#L177-L179 this regex is using native `preg_match` directly?

Regexes are usually fast.

`(...*)*` in theory can be very slow compared to `(...*+)*`.

Also the trim can be easily integrated into the regex (and as the linked PHP CS Fixer already does).

And the main slowdown seems to be comming from Nette. As long as the regex is valid, no need to check for anything else than the return value. This is what PHPStan already checks and check well.

**Comment by ondrejmirtes** [comment]
Awesome, thank you!
